tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,711 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package chunks
+
+import (
+	"bufio"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+	tsdb_errors "github.com/prometheus/prometheus/tsdb/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+// Segment header fields constants.
+const (
+	// MagicHeadChunks is 4 bytes at the head of a segment file.
+	MagicHeadChunks = 0x0130BC91
+
+	headChunksFormatV1 = 1
+	inBufferShards     = 16
+	writeBufferSize    = 2 * 1024 * 1024
+)
+
+var (
+	// DefaultHeadChunkSegmentTime is the default chunks segment time range.
+	// Assuming a general scrape interval of 15s, a chunk with 120 samples would
+	// be cut every 30m, so anything <30m will cause lots of empty files. And keeping
+	// it exactly 30m also has a chance of having empty files as its near that border.
+	// Hence keeping it a little more than 30m, i.e. 40m.
+	DefaultHeadChunkSegmentTime = 40 * int64(time.Minute/time.Millisecond)
+
+	// ErrHeadReadWriterClosed returned by any method indicates
+	// that the HeadReadWriter was closed.
+	ErrHeadReadWriterClosed = errors.New("HeadReadWriter closed")
+)
+
+// HeadReadWriter is for writing the Head block chunks to the disk
+// and access chunks via an mmapped file.
+type HeadReadWriter struct {
```

Comment:
Can we name it in more explicit way? It's actually very specific head for chunks. So maybe just `HeadChunks`?

Reply:
SGTM

Reply:
But I see this can be used by anything and not just Head. How about `LiveChunkMmapper`? :P Or `LiveChunkReadWriter`

Reply:
Making it `LiveChunkDiskMapper`